File: lib/quiver.ts
import { getCache, setCache, incrementCache } from "./cache"
import {
  getRecentTradesFromDb,
  getTradesByBioguide,
  upsertTrades,
  writeBack,
  type DbTrade,
} from "./db"

export type RawCongressTrade = {
  Bioguide?: string
  Ticker?: string
  Transaction?: string
  Range?: string
  ReportDate?: string
  Representative?: string
  Party?: string
  Chamber?: string
  UniqueID?: string | number
  AssetDescription?: string
  AssetType?: string
  Date?: string
  Traded?: string
  Trade_Size_USD?: number | string
}

const QUIVER_API_HEADERS = {
  Accept: "application/json",
  "User-Agent": "CongressMoneyMan/1.0",
}

const CIRCUIT_BREAKER_KEY = "quiver:circuit-breaker"
const CIRCUIT_BREAKER_WINDOW_SECONDS = 60
const CIRCUIT_BREAKER_THRESHOLD = 10
const MAX_RETRIES = 3
const RETRY_BASE_DELAYS_MS = [500, 1500, 4500]

export class QuiverCircuitOpenError extends Error {
  constructor() {
    super("Quiver circuit breaker is open — too many recent upstream failures")
    this.name = "QuiverCircuitOpenError"
  }
}

const sleep = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms))

async function recordQuiverFailure(): Promise<void> {
  const count = await incrementCache(
    CIRCUIT_BREAKER_KEY,
    CIRCUIT_BREAKER_WINDOW_SECONDS
  )
  if (count !== null && count > CIRCUIT_BREAKER_THRESHOLD) {
    console.error(
      `[quiver] circuit breaker tripped: ${count} failures in ${CIRCUIT_BREAKER_WINDOW_SECONDS}s`
    )
  }
}

async function isCircuitOpen(): Promise<boolean> {
  const count = await getCache<number>(CIRCUIT_BREAKER_KEY)
  return typeof count === "number" && count > CIRCUIT_BREAKER_THRESHOLD
}

export async function fetchQuiverWithRetry(
  url: string,
  init: RequestInit & { next?: { revalidate?: number } }
): Promise<Response> {
  if (await isCircuitOpen()) {
    throw new QuiverCircuitOpenError()
  }

  let lastError: unknown

  for (let attempt = 0; attempt <= MAX_RETRIES; attempt++) {
    try {
      const response = await fetch(url, init)

      if (response.ok) return response

      // Retry only on rate-limiting / transient server errors.
      if (response.status === 429 || response.status >= 500) {
        await recordQuiverFailure()
        lastError = new Error(`Quiver responded ${response.status}`)
        console.warn(
          `[quiver] ${response.status} on ${url} (attempt ${attempt + 1}/${MAX_RETRIES + 1})`
        )
      } else {
        return response
      }
    } catch (error) {
      await recordQuiverFailure()
      lastError = error
      console.warn(
        `[quiver] network error on ${url} (attempt ${attempt + 1}/${MAX_RETRIES + 1}):`,
        error
      )
    }

    if (attempt < MAX_RETRIES) {
      const base = RETRY_BASE_DELAYS_MS[attempt]
      const jitter = Math.floor(Math.random() * 250)
      await sleep(base + jitter)
    }
  }

  throw lastError instanceof Error
    ? lastError
    : new Error("Quiver request failed after retries")
}

const CONGRESS_TRADES_KEY = "congress-trades"
const CONGRESS_TRADES_TTL_SECONDS = 15 * 60

export function tradeToDbRow(trade: RawCongressTrade): DbTrade | null {
  const tradeId = trade.UniqueID != null ? String(trade.UniqueID) : null
  if (!tradeId || !trade.Bioguide) return null

  const rawSize = trade.Trade_Size_USD
  const size =
    typeof rawSize === "number"
      ? rawSize
      : rawSize != null && rawSize !== ""
      ? Number(rawSize)
      : null

  return {
    trade_id: tradeId,
    bioguide_id: trade.Bioguide,
    member_name: trade.Representative ?? null,
    party: trade.Party ?? null,
    chamber: trade.Chamber ?? null,
    ticker: trade.Ticker ?? null,
    asset_name: trade.AssetDescription ?? null,
    asset_type: trade.AssetType ?? null,
    transaction_type: trade.Transaction ?? null,
    transaction_date: trade.Date ?? null,
    traded: trade.Traded ?? null,
    range_text: trade.Range ?? null,
    trade_size_usd: size != null && Number.isFinite(size) ? size : null,
    filed_at: trade.ReportDate ?? null,
  }
}

function dbRowToTrade(row: DbTrade): RawCongressTrade {
  return {
    Bioguide: row.bioguide_id,
    Ticker: row.ticker ?? undefined,
    Transaction: row.transaction_type ?? undefined,
    Range: row.range_text ?? undefined,
    ReportDate: row.filed_at ?? undefined,
    Representative: row.member_name ?? undefined,
    Party: row.party ?? undefined,
    Chamber: row.chamber ?? undefined,
    UniqueID: row.trade_id,
    AssetDescription: row.asset_name ?? undefined,
    AssetType: row.asset_type ?? undefined,
    Date: row.transaction_date ?? undefined,
    Traded: row.traded ?? undefined,
    Trade_Size_USD: row.trade_size_usd ?? undefined,
  }
}

type RawQuiverApiTrade = {
  Representative?: string
  BioGuideID?: string
  ReportDate?: string
  TransactionDate?: string
  Ticker?: string
  Transaction?: string
  Range?: string
  House?: string
  Amount?: string | number
  Party?: string
  TickerType?: string
  Description?: string | null
}

function syntheticTradeId(raw: RawQuiverApiTrade): string {
  return [
    raw.BioGuideID ?? "",
    raw.TransactionDate ?? "",
    raw.Ticker ?? "",
    raw.Transaction ?? "",
    raw.Amount ?? "",
    raw.ReportDate ?? "",
  ].join("|")
}

function normalizeQuiverTrade(raw: RawQuiverApiTrade): RawCongressTrade {
  return {
    UniqueID: syntheticTradeId(raw),
    Bioguide: raw.BioGuideID,
    Representative: raw.Representative,
    Party: raw.Party,
    Chamber: raw.House,
    Ticker: raw.Ticker,
    AssetDescription: raw.Description ?? undefined,
    AssetType: raw.TickerType,
    Transaction: raw.Transaction,
    Date: raw.TransactionDate,
    Range: raw.Range,
    Trade_Size_USD: raw.Amount,
    ReportDate: raw.ReportDate,
  }
}

export async function fetchAllCongressTrades(
  apiKey: string,
  opts?: { forceRefresh?: boolean }
): Promise<RawCongressTrade[]> {
  const force = opts?.forceRefresh ?? false

  if (!force) {
    const cached = await getCache<RawCongressTrade[]>(CONGRESS_TRADES_KEY)
    if (cached) return cached
    const stored = await getRecentTradesFromDb()
    if (stored.length > 0) {
      const trades = stored.map(dbRowToTrade)
      await setCache(CONGRESS_TRADES_KEY, trades, CONGRESS_TRADES_TTL_SECONDS)
      return trades
    }
  }

  const response = await fetchQuiverWithRetry(
    "https://api.quiverquant.com/beta/live/congresstrading",
    {
      headers: {
        ...QUIVER_API_HEADERS,
        Authorization: `Bearer ${apiKey}`,
      },
      next: { revalidate: 900 },
    }
  )

  if (!response.ok) {
    const body = await response.text().catch(() => "")
    throw new Error(`Quiver API error ${response.status}: ${body.slice(0, 200)}`)
  }

  const data = await response.json()
  if (!Array.isArray(data)) throw new Error("Unexpected Quiver response format")

  const normalized = (data as RawQuiverApiTrade[]).map(normalizeQuiverTrade)
  const trades = [...new Map(normalized.map((t) => [t.UniqueID, t])).values()]

  await setCache(CONGRESS_TRADES_KEY, trades, CONGRESS_TRADES_TTL_SECONDS)

  if (!force) {
    const rows = trades
      .map(tradeToDbRow)
      .filter((row): row is DbTrade => row !== null)
    writeBack(() => upsertTrades(rows))
  }

  return trades
}

type RawQuiverBulkTrade = {
  Ticker?: string
  TickerType?: string
  Company?: string | null
  Traded?: string
  Transaction?: string
  Trade_Size_USD?: string | number
  Description?: string | null
  Name?: string
  BioGuideID?: string
  Filed?: string
  Party?: string
  Chamber?: string
}

function normalizeBulkTrade(raw: RawQuiverBulkTrade): RawCongressTrade {
  // Reuse the live-feed id logic so the same disclosure dedupes across feeds.
  const id = syntheticTradeId({
    BioGuideID: raw.BioGuideID,
    TransactionDate: raw.Traded,
    Ticker: raw.Ticker,
    Transaction: raw.Transaction,
    Amount: raw.Trade_Size_USD,
    ReportDate: raw.Filed,
  })
  return {
    UniqueID: id,
    Bioguide: raw.BioGuideID,
    Representative: raw.Name,
    Party: raw.Party,
    Chamber: raw.Chamber,
    Ticker: raw.Ticker,
    AssetDescription: raw.Description ?? raw.Company ?? undefined,
    AssetType: raw.TickerType,
    Transaction: raw.Transaction,
    Date: raw.Traded,
    Traded: raw.Traded,
    Range: undefined,
    Trade_Size_USD: raw.Trade_Size_USD,
    ReportDate: raw.Filed,
  }
}
const BULK_PAGE_SIZE = 1000
const BULK_MAX_PAGES = 200

async function fetchBulkPage(
  apiKey: string,
  page: number
): Promise<RawQuiverBulkTrade[] | null> {
  const url = `https://api.quiverquant.com/beta/bulk/congresstrading?page=${page}&page_size=${BULK_PAGE_SIZE}`
  for (let attempt = 0; attempt <= MAX_RETRIES; attempt++) {
    const res = await fetch(url, {
      headers: { ...QUIVER_API_HEADERS, Authorization: `Bearer ${apiKey}` },
    })
    if (res.ok) {
      const data = await res.json()
      return Array.isArray(data) ? (data as RawQuiverBulkTrade[]) : null
    }
    const transient = res.status === 429 || res.status >= 500
    if (transient && attempt < MAX_RETRIES) {
      console.warn(`[quiver] bulk page ${page} -> ${res.status}, retry ${attempt + 1}`)
      await sleep(RETRY_BASE_DELAYS_MS[attempt] + Math.floor(Math.random() * 250))
      continue
    }
    return null
  }
  return null
}

export async function fetchBulkCongressTrades(
  apiKey: string
): Promise<RawCongressTrade[]> {
  const all: RawCongressTrade[] = []
  for (let page = 1; page <= BULK_MAX_PAGES; page++) {
    const data = await fetchBulkPage(apiKey, page)
    if (!data || data.length === 0) break
    for (const raw of data) all.push(normalizeBulkTrade(raw))
    if (data.length < BULK_PAGE_SIZE) break
  }
  return [...new Map(all.map((t) => [t.UniqueID, t])).values()]
}

export async function fetchMemberCongressTrades(
  bioguideId: string,
  apiKey: string
): Promise<RawCongressTrade[]> {
  const rows = await getTradesByBioguide(bioguideId)
  if (rows.length > 0) return rows.map(dbRowToTrade)
  const all = await fetchAllCongressTrades(apiKey)
  return all.filter((t) => t.Bioguide === bioguideId)
}

export function classifyTransaction(value?: string): "buy" | "sell" | "other" {
  const v = value?.toLowerCase() ?? ""
  if (v.includes("purchase") || v.includes("buy")) return "buy"
  if (v.includes("sale") || v.includes("sell") || v.includes("sold")) return "sell"
  return "other"
}

function tradeTime(trade: RawCongressTrade): number {
  return Date.parse(trade.Date ?? "")
}
export function findMatchingSale(
  all: RawCongressTrade[],
  purchase: RawCongressTrade
): RawCongressTrade | null {
  const boughtAt = tradeTime(purchase)
  if (!Number.isFinite(boughtAt)) return null

  return (
    all
      .filter(
        (t) =>
          t.Bioguide === purchase.Bioguide &&
          t.Ticker === purchase.Ticker &&
          classifyTransaction(t.Transaction) === "sell" &&
          Number.isFinite(tradeTime(t)) &&
          tradeTime(t) >= boughtAt
      )
      .sort((a, b) => tradeTime(a) - tradeTime(b))[0] ?? null
  )
}

export function findMatchingPurchase(
  all: RawCongressTrade[],
  sale: RawCongressTrade
): RawCongressTrade | null {
  const soldAt = tradeTime(sale)
  if (!Number.isFinite(soldAt)) return null

  return (
    all
      .filter(
        (t) =>
          t.Bioguide === sale.Bioguide &&
          t.Ticker === sale.Ticker &&
          classifyTransaction(t.Transaction) === "buy" &&
          Number.isFinite(tradeTime(t)) &&
          tradeTime(t) <= soldAt
      )
      .sort((a, b) => tradeTime(b) - tradeTime(a))[0] ?? null
  )
}

export function parseTradeRange(range?: string): { low: number; high: number } | null {
  if (!range) return null
  const nums = range.replace(/,/g, "").match(/\d+(?:\.\d+)?/g)
  if (!nums || nums.length === 0) return null
  const low = Number(nums[0])
  const high = nums.length > 1 ? Number(nums[1]) : low
  if (!Number.isFinite(low) || !Number.isFinite(high)) return null
  return { low, high }
}

export function formatTradeRange(lowerBound: number): string {
  const ranges: [number, number][] = [
    [1, 1000],
    [1001, 15000],
    [15001, 50000],
    [50001, 100000],
    [100001, 250000],
    [250001, 500000],
    [500001, 1000000],
    [1000001, 5000000],
    [5000001, 25000000],
    [25000001, 50000000],
  ]

  for (const [min, max] of ranges) {
    if (lowerBound === min) return `$${min.toLocaleString()} – $${max.toLocaleString()}`
  }

  if (lowerBound >= 50000001) return `$${lowerBound.toLocaleString()}+`
  return `$${lowerBound.toLocaleString()}`
}

File: backend/app/clients/quiver.py
"""Quiver Quantitative client (port of lib/quiver.ts).

Trades are plain dicts with Quiver's raw keys (`Bioguide`, `Ticker`,
`Transaction`, `UniqueID`, ...) so cached values stay interchangeable with the
Next.js backend.
"""

import asyncio
import logging
import random
import re

from ..core.cache import get_cache, increment_cache, set_cache
from ..core.db import get_recent_trades_from_db, get_trades_by_bioguide, upsert_trades, write_back
from ..core.http import shared_client
from ..core.util import parse_ms

logger = logging.getLogger("quiver")

QUIVER_API_HEADERS = {
    "Accept": "application/json",
    "User-Agent": "CongressMoneyMan/1.0",
}

CIRCUIT_BREAKER_KEY = "quiver:circuit-breaker"
CIRCUIT_BREAKER_WINDOW_SECONDS = 60
CIRCUIT_BREAKER_THRESHOLD = 10
MAX_RETRIES = 3
RETRY_BASE_DELAYS_MS = [500, 1500, 4500]

CONGRESS_TRADES_KEY = "congress-trades"
CONGRESS_TRADES_TTL_SECONDS = 15 * 60


class QuiverCircuitOpenError(Exception):
    def __init__(self) -> None:
        super().__init__(
            "Quiver circuit breaker is open — too many recent upstream failures"
        )


async def _record_quiver_failure() -> None:
    count = await increment_cache(CIRCUIT_BREAKER_KEY, CIRCUIT_BREAKER_WINDOW_SECONDS)
    if count is not None and count > CIRCUIT_BREAKER_THRESHOLD:
        logger.error(
            "circuit breaker tripped: %s failures in %ss",
            count,
            CIRCUIT_BREAKER_WINDOW_SECONDS,
        )


async def _is_circuit_open() -> bool:
    count = await get_cache(CIRCUIT_BREAKER_KEY)
    return isinstance(count, (int, float)) and count > CIRCUIT_BREAKER_THRESHOLD


async def fetch_quiver_with_retry(url: str, headers: dict[str, str]):
    if await _is_circuit_open():
        raise QuiverCircuitOpenError()

    last_error: Exception | None = None

    for attempt in range(MAX_RETRIES + 1):
        try:
            response = await shared_client().get(url, headers=headers)

            if response.status_code < 400:
                return response

            # Retry only on rate-limiting / transient server errors.
            if response.status_code == 429 or response.status_code >= 500:
                await _record_quiver_failure()
                last_error = RuntimeError(f"Quiver responded {response.status_code}")
                logger.warning(
                    "%s on %s (attempt %s/%s)",
                    response.status_code,
                    url,
                    attempt + 1,
                    MAX_RETRIES + 1,
                )
            else:
                return response
        except Exception as error:
            await _record_quiver_failure()
            last_error = error
            logger.warning(
                "network error on %s (attempt %s/%s): %s",
                url,
                attempt + 1,
                MAX_RETRIES + 1,
                error,
            )

        if attempt < MAX_RETRIES:
            base = RETRY_BASE_DELAYS_MS[attempt]
            await asyncio.sleep((base + random.randint(0, 249)) / 1000)

    raise last_error or RuntimeError("Quiver request failed after retries")


def trade_to_db_row(trade: dict) -> dict | None:
    trade_id = js_str(trade["UniqueID"]) if trade.get("UniqueID") is not None else None
    if not trade_id or not trade.get("Bioguide"):
        return None

    raw_size = trade.get("Trade_Size_USD")
    size: float | None = None
    if isinstance(raw_size, (int, float)):
        size = float(raw_size)
    elif raw_size not in (None, ""):
        try:
            size = float(raw_size)
        except (TypeError, ValueError):
            size = None

    return {
        "trade_id": trade_id,
        "bioguide_id": trade["Bioguide"],
        "member_name": trade.get("Representative"),
        "party": trade.get("Party"),
        "chamber": trade.get("Chamber"),
        "ticker": trade.get("Ticker"),
        "asset_name": trade.get("AssetDescription"),
        "asset_type": trade.get("AssetType"),
        "transaction_type": trade.get("Transaction"),
        "transaction_date": trade.get("Date"),
        "traded": trade.get("Traded"),
        "range_text": trade.get("Range"),
        "trade_size_usd": size,
        "filed_at": trade.get("ReportDate"),
    }


def db_row_to_trade(row: dict) -> dict:
    return {
        "Bioguide": row.get("bioguide_id"),
        "Ticker": row.get("ticker"),
        "Transaction": row.get("transaction_type"),
        "Range": row.get("range_text"),
        "ReportDate": row.get("filed_at"),
        "Representative": row.get("member_name"),
        "Party": row.get("party"),
        "Chamber": row.get("chamber"),
        "UniqueID": row.get("trade_id"),
        "AssetDescription": row.get("asset_name"),
        "AssetType": row.get("asset_type"),
        "Date": row.get("transaction_date"),
        "Traded": row.get("traded"),
        "Trade_Size_USD": row.get("trade_size_usd"),
    }


def js_str(value) -> str:
    """Stringify like JS String(): integral floats have no decimal point, so
    synthetic ids match the ones the TS backend already persisted."""
    if value is None:
        return ""
    if isinstance(value, float) and value.is_integer():
        return str(int(value))
    return str(value)


def _synthetic_trade_id(raw: dict) -> str:
    return "|".join(
        js_str(raw.get(key))
        for key in (
            "BioGuideID",
            "TransactionDate",
            "Ticker",
            "Transaction",
            "Amount",
            "ReportDate",
        )
    )


def _normalize_quiver_trade(raw: dict) -> dict:
    return {
        "UniqueID": _synthetic_trade_id(raw),
        "Bioguide": raw.get("BioGuideID"),
        "Representative": raw.get("Representative"),
        "Party": raw.get("Party"),
        "Chamber": raw.get("House"),
        "Ticker": raw.get("Ticker"),
        "AssetDescription": raw.get("Description"),
        "AssetType": raw.get("TickerType"),
        "Transaction": raw.get("Transaction"),
        "Date": raw.get("TransactionDate"),
        "Range": raw.get("Range"),
        "Trade_Size_USD": raw.get("Amount"),
        "ReportDate": raw.get("ReportDate"),
    }


async def fetch_all_congress_trades(
    api_key: str, force_refresh: bool = False
) -> list[dict]:
    if not force_refresh:
        cached = await get_cache(CONGRESS_TRADES_KEY)
        if cached:
            return cached
        stored = await get_recent_trades_from_db()
        if stored:
            trades = [db_row_to_trade(row) for row in stored]
            await set_cache(CONGRESS_TRADES_KEY, trades, CONGRESS_TRADES_TTL_SECONDS)
            return trades

    response = await fetch_quiver_with_retry(
        "https://api.quiverquant.com/beta/live/congresstrading",
        {**QUIVER_API_HEADERS, "Authorization": f"Bearer {api_key}"},
    )

    if response.status_code >= 400:
        raise RuntimeError(
            f"Quiver API error {response.status_code}: {response.text[:200]}"
        )

    data = response.json()
    if not isinstance(data, list):
        raise RuntimeError("Unexpected Quiver response format")

    normalized = [_normalize_quiver_trade(raw) for raw in data]
    trades = list({t["UniqueID"]: t for t in normalized}.values())

    await set_cache(CONGRESS_TRADES_KEY, trades, CONGRESS_TRADES_TTL_SECONDS)

    if not force_refresh:
        rows = [row for row in (trade_to_db_row(t) for t in trades) if row is not None]
        write_back(upsert_trades(rows))

    return trades


def _normalize_bulk_trade(raw: dict) -> dict:
    # Reuse the live-feed id logic so the same disclosure dedupes across feeds.
    trade_id = _synthetic_trade_id(
        {
            "BioGuideID": raw.get("BioGuideID"),
            "TransactionDate": raw.get("Traded"),
            "Ticker": raw.get("Ticker"),
            "Transaction": raw.get("Transaction"),
            "Amount": raw.get("Trade_Size_USD"),
            "ReportDate": raw.get("Filed"),
        }
    )
    description = raw.get("Description")
    if description is None:
        description = raw.get("Company")
    return {
        "UniqueID": trade_id,
        "Bioguide": raw.get("BioGuideID"),
        "Representative": raw.get("Name"),
        "Party": raw.get("Party"),
        "Chamber": raw.get("Chamber"),
        "Ticker": raw.get("Ticker"),
        "AssetDescription": description,
        "AssetType": raw.get("TickerType"),
        "Transaction": raw.get("Transaction"),
        "Date": raw.get("Traded"),
        "Traded": raw.get("Traded"),
        "Range": None,
        "Trade_Size_USD": raw.get("Trade_Size_USD"),
        "ReportDate": raw.get("Filed"),
    }


BULK_PAGE_SIZE = 1000
BULK_MAX_PAGES = 200


async def _fetch_bulk_page(api_key: str, page: int) -> list[dict] | None:
    url = (
        "https://api.quiverquant.com/beta/bulk/congresstrading"
        f"?page={page}&page_size={BULK_PAGE_SIZE}"
    )
    for attempt in range(MAX_RETRIES + 1):
        response = await shared_client().get(
            url, headers={**QUIVER_API_HEADERS, "Authorization": f"Bearer {api_key}"}
        )
        if response.status_code < 400:
            data = response.json()
            return data if isinstance(data, list) else None
        transient = response.status_code == 429 or response.status_code >= 500
        if transient and attempt < MAX_RETRIES:
            logger.warning(
                "bulk page %s -> %s, retry %s", page, response.status_code, attempt + 1
            )
            await asyncio.sleep(
                (RETRY_BASE_DELAYS_MS[attempt] + random.randint(0, 249)) / 1000
            )
            continue
        return None
    return None


async def fetch_bulk_congress_trades(api_key: str) -> list[dict]:
    all_trades: list[dict] = []
    for page in range(1, BULK_MAX_PAGES + 1):
        data = await _fetch_bulk_page(api_key, page)
        if not data:
            break
        all_trades.extend(_normalize_bulk_trade(raw) for raw in data)
        if len(data) < BULK_PAGE_SIZE:
            break
    return list({t["UniqueID"]: t for t in all_trades}.values())


async def fetch_member_congress_trades(bioguide_id: str, api_key: str) -> list[dict]:
    rows = await get_trades_by_bioguide(bioguide_id)
    if rows:
        return [db_row_to_trade(row) for row in rows]
    all_trades = await fetch_all_congress_trades(api_key)
    return [t for t in all_trades if t.get("Bioguide") == bioguide_id]


def classify_transaction(value: str | None) -> str:
    v = (value or "").lower()
    if "purchase" in v or "buy" in v:
        return "buy"
    if "sale" in v or "sell" in v or "sold" in v:
        return "sell"
    return "other"


def _trade_time(trade: dict) -> float | None:
    return parse_ms(trade.get("Date"))


def find_matching_sale(all_trades: list[dict], purchase: dict) -> dict | None:
    bought_at = _trade_time(purchase)
    if bought_at is None:
        return None

    candidates = [
        t
        for t in all_trades
        if t.get("Bioguide") == purchase.get("Bioguide")
        and t.get("Ticker") == purchase.get("Ticker")
        and classify_transaction(t.get("Transaction")) == "sell"
        and _trade_time(t) is not None
        and _trade_time(t) >= bought_at
    ]
    candidates.sort(key=_trade_time)
    return candidates[0] if candidates else None


def find_matching_purchase(all_trades: list[dict], sale: dict) -> dict | None:
    sold_at = _trade_time(sale)
    if sold_at is None:
        return None

    candidates = [
        t
        for t in all_trades
        if t.get("Bioguide") == sale.get("Bioguide")
        and t.get("Ticker") == sale.get("Ticker")
        and classify_transaction(t.get("Transaction")) == "buy"
        and _trade_time(t) is not None
        and _trade_time(t) <= sold_at
    ]
    candidates.sort(key=_trade_time, reverse=True)
    return candidates[0] if candidates else None


def parse_trade_range(range_text: str | None) -> dict | None:
    if not range_text:
        return None
    nums = re.findall(r"\d+(?:\.\d+)?", range_text.replace(",", ""))
    if not nums:
        return None
    low = float(nums[0])
    high = float(nums[1]) if len(nums) > 1 else low
    return {"low": low, "high": high}


def format_trade_range(lower_bound: float) -> str:
    ranges = [
        (1, 1000),
        (1001, 15000),
        (15001, 50000),
        (50001, 100000),
        (100001, 250000),
        (250001, 500000),
        (500001, 1000000),
        (1000001, 5000000),
        (5000001, 25000000),
        (25000001, 50000000),
    ]

    for minimum, maximum in ranges:
        if lower_bound == minimum:
            return f"${minimum:,} – ${maximum:,}"

    if lower_bound >= 50000001:
        return f"${lower_bound:,.0f}+"
    return f"${lower_bound:,.0f}"
